pkt-io/src/nf.rs

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+//! Local packet I/O for gateway
+
+#![allow(unused)]
+
+use crossbeam::queue::ArrayQueue;
+use net::packet::Packet;
+use net::{buffer::PacketBufferMut, packet::DoneReason};
+use pipeline::NetworkFunction;
+use std::sync::Arc;
+use tracectl::trace_target;
+use tracing::trace;
+
+const PKT_IO: &str = "pkt-io";
+trace_target!(PKT_IO, LevelFilter::TRACE, &["pipeline"]);
+
+// Type for a network function that performs two basic functions:
+//    1) ability to inject packets into a pipeline (e.g. to transmit them).
+//    2) ability to pull packets out of a pipeline (e.g. "punt" for local consumption).
+//
+// The above is achieved using two queues. One (inject) for the packets to be injected into the
+// pipeline, which may be populated by other thread(s) to "send" packets to a pipeline. Another
+// one (punt) that can contain packets that have been removed from the pipeline to be processed
+// elsewhere, possibly by another thread. The two queues have a fixed capacity. They are defined
+// to hold Box<Packet>'s so that setting large capacities does not consume a large amount of memory.
+// For the same reason, the two queues are optional to accommodate for the case that only injection
+// or punting are used.
+pub struct PktIo<Buf: PacketBufferMut> {
+    injectq: Option<Arc<ArrayQueue<Box<Packet<Buf>>>>>,
```

Comment:
I _really_ don't understand this data structure choice.  At all.

Reply:
Which part? Note that the Arc is gone later